File: solutions/76.minimum-window-substring.java
/*
 * @lc app=leetcode id=76 lang=java
 *
 * [76] Minimum Window Substring
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) {
            return "";
        }
        Map<Character, Integer> tMap = new HashMap<>();
        for (Character c : t.toCharArray()) {
            tMap.merge(c, 1, Integer::sum);
        }

        int resLeft = 0;
        int resRight = -1;
        int left = 0;
        int right = 0;
        int matches = 0;
        while (right < s.length()) {
            Character charRight = s.charAt(right);
            if (tMap.containsKey(charRight)) {
                tMap.put(charRight, tMap.get(charRight) - 1);
                if (tMap.get(charRight) == 0) {
                    matches++;
                }
                while (matches == tMap.size()) {
                    if (resRight == -1 || (resRight - resLeft > right - left)) {
                        resLeft = left;
                        resRight = right;
                        if (right - left + 1 == t.length()) {
                            return s.substring(resLeft, resRight + 1);
                        }
                    }
                    Character charLeft = s.charAt(left++);
                    if (tMap.containsKey(charLeft)) {
                        if (tMap.get(charLeft) == 0) {
                            matches--;
                        }
                        tMap.put(charLeft, tMap.get(charLeft) + 1);
                    }
                }
            }
            right++;
        }

        return (resRight == -1) ? "" : s.substring(resLeft, resRight + 1);
    }
}
// @lc code=end

File: solutions/76.minimum-window-substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""

        map = {}
        for c in t:
            map[c] = map.get(c, 0) + 1

        res_left = 0
        res_right = -1
        left = 0
        right = 0
        matches = 0
        while right < len(s):
            char_add = s[right]
            if char_add in map:
                map[char_add] -= 1
                if map[char_add] == 0:
                    matches += 1

                while matches == len(map):
                    if res_right == -1 or res_right - res_left > right - left:
                        res_right = right
                        res_left = left
                        if right - left + 1 == len(t):
                            return s[res_left : res_right + 1]

                    char_del = s[left]
                    if char_del in map:
                        if map[char_del] == 0:
                            matches -= 1
                        map[char_del] += 1
                    left += 1

            right += 1

        return s[res_left : res_right + 1] if res_right > -1 else ""

# @lc code=end

